File: src/backend/ct2_python.rs
use std::path::PathBuf;
use std::process::Command;

use anyhow::{Context, Result, bail};

use crate::backend::TranscriptionBackend;
use crate::types::{BackendDescriptor, BackendKind, Transcript, TranscriptionRequest};

pub struct Ct2PythonBackend {
    worker_launcher: PathBuf,
    worker_launcher_args: Vec<String>,
    worker_script: PathBuf,
}

impl Ct2PythonBackend {
    pub fn new(
        worker_launcher: PathBuf,
        worker_launcher_args: Vec<String>,
        worker_script: PathBuf,
    ) -> Self {
        Self {
            worker_launcher,
            worker_launcher_args,
            worker_script,
        }
    }
}

impl TranscriptionBackend for Ct2PythonBackend {
    fn descriptor(&self) -> BackendDescriptor {
        super::describe_backend(BackendKind::Ct2Python)
    }

    fn transcribe(&self, request: &TranscriptionRequest) -> Result<Transcript> {
        let mut command = Command::new(&self.worker_launcher);
        command.args(&self.worker_launcher_args);
        command
            .env("UV_CACHE_DIR", ".uv-cache")
            .arg(&self.worker_script)
            .arg("--audio")
            .arg(&request.audio_path)
            .arg("--model")
            .arg(request.model.as_ct2_model_id())
            .arg("--model-repo")
            .arg(request.model.as_hf_repo());

        if let Some(language) = &request.language {
            command.arg("--language").arg(language);
        }

        if let Some(prompt) = &request.initial_prompt {
            command.arg("--initial-prompt").arg(prompt);
        }

        let output = command.output().with_context(|| {
            format!(
                "failed to run Python worker {}",
                self.worker_script.display()
            )
        })?;

        if !output.status.success() {
            bail!(
                "Python worker exited with status {}.\nstdout:\n{}\nstderr:\n{}",
                output.status,
                String::from_utf8_lossy(&output.stdout),
                String::from_utf8_lossy(&output.stderr)
            );
        }

        serde_json::from_slice(&output.stdout).with_context(|| {
            format!(
                "failed to parse worker JSON output.\nstdout:\n{}\nstderr:\n{}",
                String::from_utf8_lossy(&output.stdout),
                String::from_utf8_lossy(&output.stderr)
            )
        })
    }
}

File: src/backends/ct2_python.py
from __future__ import annotations

import json
import os
import subprocess
import sys
from dataclasses import dataclass

from ..config import MODEL_CT2_ALIAS, MODEL_HF_REPO, AppConfig
from ..types import (
    BackendDescriptor,
    BackendKind,
    Transcript,
    TranscriptionRequest,
)


@dataclass(slots=True)
class Ct2PythonBackend:
    config: AppConfig

    def descriptor(self) -> BackendDescriptor:
        from . import describe_backend

        return describe_backend(BackendKind.CT2_PYTHON)

    def build_worker_command(self, request: TranscriptionRequest) -> list[str]:
        command = [
            sys.executable,
            str(self.config.worker_script_path),
            "--audio",
            str(request.audio_path),
        ]
        if request.initial_prompt:
            command.extend(["--initial-prompt", request.initial_prompt])
        return command

    def transcribe(self, request: TranscriptionRequest) -> Transcript:
        command = self.build_worker_command(request)
        env = os.environ.copy()
        env.setdefault("UV_CACHE_DIR", str(self.config.uv_cache_dir))

        completed = subprocess.run(
            command,
            cwd=self.config.project_root,
            env=env,
            capture_output=True,
            text=True,
            check=False,
        )

        if completed.returncode != 0:
            raise RuntimeError(
                "Python worker exited with a non-zero status.\n"
                f"status: {completed.returncode}\n"
                f"stdout:\n{completed.stdout}\n"
                f"stderr:\n{completed.stderr}"
            )

        try:
            payload = json.loads(completed.stdout)
        except json.JSONDecodeError as exc:
            raise RuntimeError(
                "failed to parse worker JSON output.\n"
                f"stdout:\n{completed.stdout}\n"
                f"stderr:\n{completed.stderr}"
            ) from exc

        transcript = Transcript.from_dict(payload)
        transcript.model = MODEL_HF_REPO
        if all("Model alias:" not in note for note in transcript.notes):
            insert_at = 1 if transcript.notes else 0
            transcript.notes.insert(insert_at, f"Model alias: {MODEL_CT2_ALIAS}")
        return transcript
